api/client/proxy/proxy.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/http/httpproxy"
+)
+
+// GetProxyAddress gets the HTTP proxy address to use for a given address, if any.
+func GetProxyAddress(dialAddr string) *url.URL {
+	addrURL, err := parse(dialAddr)
+	if err != nil {
+		return nil
+	}
+	proxyFunc := httpproxy.FromEnvironment().ProxyFunc()
+	if addrURL.Scheme != "" {
+		proxyURL, err := proxyFunc(addrURL)
+		if err != nil {
+			return nil
+		}
+		return proxyURL
+	}
+
+	for _, scheme := range []string{"https", "http"} {
+		addrURL.Scheme = scheme
+		proxyURL, err := proxyFunc(addrURL)
+		if err == nil && proxyURL != nil {
+			return proxyURL
+		}
+	}
+
+	return nil
+}
+
+// parse parses a URL. If the address does not have a scheme, it will prepend "http" and try.
+func parse(addr string) (*url.URL, error) {
+	proxyurl, err := url.Parse(addr)
```

Comment:
Some things about this function are still unclear to me:
- `proxyurl` is still confusing because this function doesn't actually deal with proxy addresses now.
- In which case `addr` will not have a scheme in it?
- Shouldn't we default to `https` if the scheme is empty? That would be a more secure default, no?

Reply:
Whether or not `addr` has a scheme depends on the caller; the existing proxy tests didn't have schemes, so I'm assuming that's a common and expected scenario.

The extra `http://` is only there to help parse the url; `parse` doesn't assume http. `GetProxyAddress` handles the scheme-less URLs by checking https first, then http.